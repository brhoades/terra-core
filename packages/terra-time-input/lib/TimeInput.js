'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactIntl = require('react-intl');

var _terraI18n = require('terra-i18n');

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _reactTextMask = require('react-text-mask');

var _reactTextMask2 = _interopRequireDefault(_reactTextMask);

var _TimePipe = require('./TimePipe');

var _TimePipe2 = _interopRequireDefault(_TimePipe);

require('./TimeInput.scss');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /* eslint-disable global-require, import/no-dynamic-require */


var propTypes = {
  /**
   * A moment object to use as the default initial time.
   */
  defaultTime: _react.PropTypes.object,
  /**
  * A callback function to execute when a time value is entered.
  */
  onChange: _react.PropTypes.func,
  /**
   * Locale to format the time
   */
  locale: _react.PropTypes.string.isRequired
};

var defaultProps = {
  defaultTime: null,
  onChange: null
};

var TimeInput = function (_React$Component) {
  _inherits(TimeInput, _React$Component);

  function TimeInput(props) {
    _classCallCheck(this, TimeInput);

    var _this = _possibleConstructorReturn(this, (TimeInput.__proto__ || Object.getPrototypeOf(TimeInput)).call(this, props));

    _this.state = {
      load: false,
      locale: props.locale,
      timeMessages: require('./translations/' + props.locale + '.js'),
      defaultTime: null,
      value: null
    };

    _this.handleChange = _this.handleChange.bind(_this);
    _this.handleInputKeyDown = _this.handleInputKeyDown.bind(_this);
    return _this;
  }

  _createClass(TimeInput, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      (0, _terraI18n.i18nLoader)(this.props.locale, this.setState, this);

      this.state = {
        defaultTime: this.formattedTime(this.props.defaultTime),
        value: this.formattedTime(this.props.defaultTime)
      };
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (this.props === nextProps) return;
      (0, _terraI18n.i18nLoader)(nextProps.locale, this.setState, this);
    }
  }, {
    key: 'formattedTime',
    value: function formattedTime(timeMoment) {
      if (timeMoment) {
        debugger;
        return timeMoment.clone().locale(this.props.locale).format(_react2.default.createElement(_reactIntl.FormattedMessage, { id: 'Terra.timeInput.format' }));
      }

      return '';
    }
  }, {
    key: 'handleChange',
    value: function handleChange(event) {
      if (event.target.value === this.state.value) {
        return;
      }

      this.setState({
        value: event.target.value
      });

      if (this.props.onChange) {
        this.props.onChange(event);
      }
    }
  }, {
    key: 'handleInputKeyDown',
    value: function handleInputKeyDown(event) {
      var copy = (0, _moment2.default)(this.state.value, _react2.default.createElement(_reactIntl.FormattedMessage, { id: 'Terra.timeInput.format' }), true);

      if (!copy.isValid()) {
        return;
      }

      if (event.key === 'ArrowUp') {
        this.setState({ value: this.formattedTime(copy.add(1, 'minutes')) });
      } else if (event.key === 'ArrowDown') {
        this.setState({ value: this.formattedTime(copy.subtract(1, 'minutes')) });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      if (!this.state.load) return null;

      var _props = this.props,
          defaultTime = _props.defaultTime,
          locale = _props.locale,
          onChange = _props.onChange,
          customProps = _objectWithoutProperties(_props, ['defaultTime', 'locale', 'onChange']);

      return _react2.default.createElement(
        _terraI18n.I18nProvider,
        {
          locale: this.props.locale,
          messages: this.state.timeMessages
        },
        _react2.default.createElement(
          'div',
          { className: 'terra-TimeInput' },
          _react2.default.createElement(_reactTextMask2.default, _extends({}, customProps, {
            className: 'terra-TimeInput-input',
            type: 'text',
            value: this.state.value,
            defaultValue: this.state.defaultTime,
            onChange: this.handleChange,
            placeholder: _react2.default.createElement(_reactIntl.FormattedMessage, { id: 'Terra.timeInput.format' }),
            mask: [/[0-2]/, /[0-9]/, ':', /[0-5]/, /[0-9]/],
            keepCharPositions: true,
            placeholderChar: ' ',
            pipe: (0, _TimePipe2.default)(_react2.default.createElement(_reactIntl.FormattedMessage, { id: 'Terra.timeInput.format' })),
            onKeyDown: this.handleInputKeyDown
          }))
        )
      );
    }
  }]);

  return TimeInput;
}(_react2.default.Component);

TimeInput.propTypes = propTypes;
TimeInput.defaultProps = defaultProps;

exports.default = TimeInput;