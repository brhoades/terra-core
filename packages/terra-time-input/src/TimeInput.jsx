/* eslint-disable global-require, import/no-dynamic-require */
import React, { PropTypes } from 'react';
import { FormattedMessage } from 'react-intl';
import { I18nProvider, i18nLoader } from 'terra-i18n';
import moment from 'moment';
import TextMask from 'react-text-mask';
import TimePipe from './TimePipe';

import './TimeInput.scss';

const propTypes = {
  /**
   * A moment object to use as the default initial time.
   */
  defaultTime: PropTypes.object,
    /**
   * A callback function to execute when a time value is entered.
   */
  onChange: PropTypes.func,
  /**
   * Locale to format the time
   */
  locale: PropTypes.string.isRequired,
};

const defaultProps = {
  defaultTime: null,
  onChange: null,
};

class TimeInput extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      load: false,
      locale: props.locale,
      timeMessages: require(`./translations/${props.locale}.js`),
      defaultTime: null,
      value: null,
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleInputKeyDown = this.handleInputKeyDown.bind(this);
  }

  componentDidMount() {
    i18nLoader(this.props.locale, this.setState, this);

    this.state = {
      defaultTime: this.formattedTime(this.props.defaultTime),
      value: this.formattedTime(this.props.defaultTime),
    };
  }

  componentWillReceiveProps(nextProps) {
    if (this.props === nextProps) return;
    i18nLoader(nextProps.locale, this.setState, this);
  }

  formattedTime(timeMoment) {
    if (timeMoment) {
      debugger;
      return timeMoment.clone().locale(this.props.locale).format(<FormattedMessage id="Terra.timeInput.format" />);
    }

    return '';
  }

  handleChange(event) {
    if (event.target.value === this.state.value) {
      return;
    }

    this.setState({
      value: event.target.value,
    });

    if (this.props.onChange) {
      this.props.onChange(event);
    }
  }

  handleInputKeyDown(event) {
    const copy = moment(this.state.value, <FormattedMessage id="Terra.timeInput.format" />, true);

    if (!copy.isValid()) {
      return;
    }

    if (event.key === 'ArrowUp') {
      this.setState({ value: this.formattedTime(copy.add(1, 'minutes')) });
    } else if (event.key === 'ArrowDown') {
      this.setState({ value: this.formattedTime(copy.subtract(1, 'minutes')) });
    }
  }

  render() {
    if (!this.state.load) return null;

    const {
      defaultTime,
      locale,
      onChange,
      ...customProps
    } = this.props;

    return (
      <I18nProvider
        locale={this.props.locale}
        messages={this.state.timeMessages}
      >
        <div className="terra-TimeInput">
          <TextMask
            {...customProps}
            className="terra-TimeInput-input"
            type="text"
            value={this.state.value}
            defaultValue={this.state.defaultTime}
            onChange={this.handleChange}
            placeholder={<FormattedMessage id="Terra.timeInput.format" />}
            mask={[/[0-2]/, /[0-9]/, ':', /[0-5]/, /[0-9]/]}
            keepCharPositions
            placeholderChar=" "
            pipe={TimePipe(<FormattedMessage id="Terra.timeInput.format" />)}
            onKeyDown={this.handleInputKeyDown}
          />
        </div>
      </I18nProvider>
    );
  }
}

TimeInput.propTypes = propTypes;
TimeInput.defaultProps = defaultProps;

export default TimeInput;
