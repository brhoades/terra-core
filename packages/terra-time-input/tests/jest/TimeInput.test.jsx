import React from 'react';
import moment from 'moment';
import TimeInput from '../../src/TimeInput';

it('should render a default time input', () => {
  const timeInput = <TimeInput />;
  const wrapper = shallow(timeInput);
  expect(wrapper).toMatchSnapshot();
});

it('should render a time input with a default time', () => {
  const timeInput = <TimeInput defaultTime={moment.utc('04-01-2016', 'MM-DD-YYYY HH:mm')} />;
  const wrapper = shallow(timeInput);
  expect(wrapper).toMatchSnapshot();
});
