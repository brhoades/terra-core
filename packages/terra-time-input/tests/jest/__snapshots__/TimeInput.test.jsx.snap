exports[`test should render a default time input 1`] = `
<div
  className="terra-TimeInput">
  <MaskedInput
    className="terra-TimeInput-input"
    defaultValue=""
    keepCharPositions={true}
    mask={
      Array [
        /[0-9]/,
        /[0-9]/,
        ":",
        /[0-9]/,
        /[0-9]/,
      ]
    }
    onChange={[Function]}
    onKeyDown={[Function]}
    pipe={[Function]}
    placeholder="HH:mm"
    placeholderChar=" "
    type="text"
    value="" />
</div>
`;

exports[`test should render a time input with a default time 1`] = `
<div
  className="terra-TimeInput">
  <MaskedInput
    className="terra-TimeInput-input"
    defaultValue="00:00"
    keepCharPositions={true}
    mask={
      Array [
        /[0-9]/,
        /[0-9]/,
        ":",
        /[0-9]/,
        /[0-9]/,
      ]
    }
    onChange={[Function]}
    onKeyDown={[Function]}
    pipe={[Function]}
    placeholder="HH:mm"
    placeholderChar=" "
    type="text"
    value="00:00" />
</div>
`;
